### PR TITLE
AWS usage improvements

### PR DESCRIPTION
**Pull request by prestwich**
Adds a configuration option to use STS assumerole for AWS signers

Open questions: 
- is there a case for allowing STS to use a separate AWS region than KMS?
- is there a case for allowing the agent to access KMS keys in 2 different AWS regions?
- is there a case for [extra STS call configuration](https://rusoto.github.io/rusoto/rusoto_sts/struct.StsAssumeRoleSessionCredentialsProvider.html)?

## Motivation

Allow agent operators to use more advance AWS permissions

## Solution

- Add an optional `arn` key to the aws signer config
- if arn is present, signer instantiation attempts to call assumerole and use an STS credentials provider
- if arn is not present, signer instantiation uses a standard credentials provider

## PR Checklist

- [ ] Added Tests
- [x] Updated Documentation
- [x] Updated CHANGELOG.md for the appropriate package

### CODE REVIEW

File: nomad-core/src/signer.rs [hunk 1]
```diff
@@ -57,23 +59,54 @@ impl From<AwsSigner<'static>> for Signers {
...
+    let sts = StsClient::new(region.parse().expect("invalid region"));
+
+    let provider = StsAssumeRoleSessionCredentialsProvider::new(
+        sts,
+        arn.to_owned(),
+        "default".to_owned(),
+        None,
+        None,
+        None,
+        None,
```
Comment (ltchang2019) [inline nomad-core/src/signer.rs:82]: So we will assume operators will not configure these? Unless it becomes feature request

https://rusoto.github.io/rusoto/rusoto_sts/struct.StsAssumeRoleSessionCredentialsProvider.html

File: configuration/src/agent/signer.rs [hunk 2]
```diff
@@ -32,7 +38,8 @@ impl FromEnv for SignerConf {
         // ordering this first preferentially uses AWS if both are specified
         if let Ok(id) = std::env::var(&format!("{}_ID", prefix)) {
             if let Ok(region) = std::env::var(&format!("{}_REGION", prefix)) {
-                return Some(SignerConf::Aws { id, region });
+                let arn = std::env::var(&format!("{}_ARN", prefix)).ok();
+                return Some(SignerConf::Aws { id, region, arn });
```
[the remark this answers is not in the document]
Comment (prestwich) [reply configuration/src/agent/signer.rs:42]: where are we currently testing fromenv? I'll add a new test for the json based

[the remark this answers is not in the document]
Comment (ltchang2019) [reply configuration/src/agent/signer.rs:42]: `configuration/src/secrets.rs` and kathy settings are the main tests for building from env. We don't have to touch those though. We just need to add dummy ARN to:
- test.env: https://github.com/nomad-xyz/rust/blob/main/fixtures/env.test
- test_secrets.json: https://github.com/nomad-xyz/rust/blob/main/fixtures/test_secrets.json

File: agents/processor/src/push.rs [hunk 3]
```diff
@@ -42,11 +40,7 @@ impl std::fmt::Debug for Pusher {
 impl Pusher {
     /// Instantiate a new pusher with a region
     pub fn new(name: &str, bucket: &str, region: Region, db: NomadDB) -> Self {
-        let client = S3Client::new_with(
-            HttpClient::new().unwrap(),
-            EnvironmentProvider::with_prefix(AWS_S3_PREFIX),
-            region.clone(),
-        );
+        let client = S3Client::new(region.clone());
```
Comment (prestwich) [inline agents/processor/src/push.rs:43]: @arnaud036 note that with this change S3 and KMS now always use the same credentials. No more need for the `OPT_PROCESSOR_S3_*` env vars

**Comment by prestwich** [comment]
this branch is now based on #177. Closing this branch in favor of that PR. This work should be preserved for future use

**Pull request closed without merging**